Add --note flag to associate messages with frames

joelostblom opened the pull request:
@jmaupetit This is a continuation of #337. As mentioned there, I merged master into this branch because the rebase was a headache with so many conflicting changes (possible I'm not doing this optimally). I suspect this doesn't matter since you usually squash when merging, but it would be great if this was given priority over other PR's, since it is hard to rebase if master is updated.

I added @k4nar 's suggestions in the three first commits. Just a comment that although the plural form `--notes` is more natural for log since it is short for `short_notes`, it might be confusing since it is `--note` everywhere else. I'll let you decide what to do, if anything.

I added three commits of my own, only 4e0e4b6 is a functional change. Without it, I would not be able to edit frames since the new `note` field would be interpreted as the `updated_at` field and `watson edit` raised an error trying to convert this to a date. I have added this and a few more comments on the code also.

Close #337 #252 #93

SpotlightKid committed:
Refactor 'watson.frames':

* Made 'watson.frames.Frames' a subclass of 'collections.OrderedDict'.
* Changed order of items in 'watson.frames.Frame' to be more consistent
  with 'Frames.add()' and 'Frames.new_frame()'.

  This changes the on-disk JSON representation of the frames, i.e. the
  format of the '<app dir>/watson/frames' file. The code handles loading
  the old format automatically and on saving the file gets written in
  the new format.
* Removed ability to look up frame by index with dict look-up syntax
  and provide 'Frames.get_by_index()' method as replacement.
* Removed ability to delete frame by index (was not used aywhere).
* Removed ability to get Frames column values with dict look-up syntax
  and provide 'Frames.get_column()' generator as replacement.
* Iterating over 'Watson.frames' yields frame IDs instead of 'Frame'
  instances. Where the latter is desired, the 'values()' method must be
  used.
* Adding a frame by assigning a 'Frame' instance to a key of a `Frames`,
  makes a copy of the 'Frame' instance with the given key as the frame ID.
* Made 'Frame.filter()' a generator.
* Adapted code using 'Frame' and 'Frames' to these changes.
* Adapted tests and test data files.
+ Command line interface remains unchanged.

SpotlightKid committed:
Fix flake8 complaint by refactoring boolean expr

SpotlightKid committed:
Optimize index access to Frames and add tests for 'get_by_index'

SpotlightKid committed:
Add log message to frames and allow setting it with stop or edit command

SpotlightKid committed:
Update merge command to work with new frame format

SpotlightKid committed:
Remove uneccessary assignment since both names already refer to the same object

Signed-off-by: Christopher Arndt <chris@chrisarndt.de>

SpotlightKid committed:
Assigning a Frames instance to watson.frames should work with new Frames format too; Clarify docstring on passing frames to construcutor as well

Signed-off-by: Christopher Arndt <chris@chrisarndt.de>

SpotlightKid committed:
Make sure passing a sequence of frames works with tuple and list items

Signed-off-by: Christopher Arndt <chris@chrisarndt.de>

SpotlightKid committed:
Add more tests for setting watson.frames

SpotlightKid committed:
Add tests for log messages

Signed-off-by: Christopher Arndt <chris@chrisarndt.de>

SpotlightKid committed:
Add options for stop command to bash completion config

Signed-off-by: Christopher Arndt <chris@chrisarndt.de>

SpotlightKid committed:
(Re-)add support for looking up frames by ID prefix

Signed-off-by: Christopher Arndt <chris@chrisarndt.de>

SpotlightKid committed:
When looking up a frame by ID prefix, iterate over keys in reverse order for better perfomance in most likely case

Signed-off-by: Christopher Arndt <chris@chrisarndt.de>

SpotlightKid committed:
Add script to convert between old and new frame format (for development)

Signed-off-by: Christopher Arndt <chris@chrisarndt.de>

SpotlightKid committed:
Update documentation for stop command

Signed-off-by: Christopher Arndt <chris@chrisarndt.de>

SpotlightKid committed:
Enhanced stop command help

Signed-off-by: Christopher Arndt <chris@chrisarndt.de>

SpotlightKid committed:
Updated user guide

Signed-off-by: Christopher Arndt <chris@chrisarndt.de>

jnsebgosselin committed:
Revert "Add script to convert between old and new frame format (for development)"

This reverts commit c7c3c84f8756a9861f7ec520db8906f273e83a25.

jnsebgosselin committed:
Revert "When looking up a frame by ID prefix, iterate over keys in reverse order for better perfomance in most likely case"

This reverts commit 8d2206ff3f1eea4512df3896ff57e01549158998.

jnsebgosselin committed:
Revert "(Re-)add support for looking up frames by ID prefix"

This reverts commit c956089e176a0c6792be4ebe927e5dbc605e85f7.

jnsebgosselin committed:
Revert "Add more tests for setting watson.frames"

This reverts commit 73c25931b40b6ebca294219eb46354153e585485.

jnsebgosselin committed:
Revert "Make sure passing a sequence of frames works with tuple and list items"

This reverts commit 714b58159b89f56cf301d5ac429c7c6b32a16ba7.

jnsebgosselin committed:
Revert "Assigning a Frames instance to watson.frames should work with new Frames format too; Clarify docstring on passing frames to construcutor as well"

This reverts commit 553e34a73799d5e94941a84da94337d244ddaf53.

jnsebgosselin committed:
Revert "Remove uneccessary assignment since both names already refer to the same object"

This reverts commit 8ed29c8f82ce7734796d1997736ad75eedf5a5f1.

jnsebgosselin committed:
Merge branch 'feature/refactor-frames' into feature/log-message

Signed-off-by: Christopher Arndt <chris@chrisarndt.de>

Merge branch 'feature/log-message' of github.com:SpotlightKid/Watson
into feature/log-message

Signed-off-by: Christopher Arndt
<chris@chrisarndt.de>

jnsebgosselin committed:
Revert "Optimize index access to Frames and add tests for 'get_by_index'"

This reverts commit a1a905a336fd05d4c99d613509e001c602ea73a3.

jnsebgosselin committed:
Revert "Fix flake8 complaint by refactoring boolean expr"

This reverts commit 92b8733b66c41ef4d8bb5687994d979b28c1d6d3.

jnsebgosselin committed:
Revert "Refactor 'watson.frames'"

jnsebgosselin committed:
Add back a missing ) deleted by mistake

jnsebgosselin committed:
Merge remote-tracking branch 'refs/remotes/TailorDev/master' into add_log_message

# Conflicts:
#	tests/test_watson.py
#	watson.completion
#	watson/cli.py

jnsebgosselin committed:
Merge remote-tracking branch 'refs/remotes/TailorDev/master' into add_log_message

jnsebgosselin committed:
Merge remote-tracking branch 'refs/remotes/TailorDev/master' into add_log_message

jnsebgosselin committed:
Merge remote-tracking branch 'refs/remotes/TailorDev/master' into add_log_message

jnsebgosselin committed:
Fix test_frames_with_message

jnsebgosselin committed:
Fix test test_frames_without_message

jnsebgosselin committed:
Fix test_stop_started_project_without_message

jnsebgosselin committed:
Fix test_stop_started_project_with_message

jnsebgosselin committed:
Remove blank spaces

jnsebgosselin committed:
Fix test_save_empty_current

prat0088 committed:
merge log message branch and master

prat0088 committed:
fix: flake8 error

prat0088 committed:
add venv to .gitignore

prat0088 committed:
fix: message wasn't saving; move some logic from cli to watson

prat0088 committed:
write message to csv, json

prat0088 committed:
update stop message test

prat0088 committed:
merge upstream master

prat0088 committed:
properly handle messages in edit; pass message to start

prat0088 committed:
restructure log print loop

prat0088 committed:
refactor

prat0088 committed:
include messages when building reports

prat0088 committed:
clean up flake8 errors

prat0088 committed:
add tests for report log messages

prat0088 committed:
wip: report/aggregate messages

prat0088 committed:
change -m/--message to -n/--note

-m and -M are currently used by some commands.  -n and -N aren't.  This diverges from the git -m/--message naming we were copying, but that doesn't bother me too much.  The parameter behaves slightly different in watson, so it might even be better that we don't reuse a commonly used parameter name in another tool, as not to cause confusion

prat0088 committed:
fix: aggregate/report -n params not detected as flags on Windows

joelostblom committed:
Merge branch 'master' into note-flag

joelostblom committed:
Remove comment

joelostblom committed:
Print note and warning message when overwriting

joelostblom committed:
Always pass id to new_frame

Without this, the note is sometimes interpreted as `updated_at` which throws an error

joelostblom committed:
Align rows of multiline notes

joelostblom committed:
Change note flag to plural for log only

joelostblom committed:
Join unnecesarily split rows

joelostblom committed:
Realign to fix flake8 test

joelostblom committed:
Remove extra venv from gitignore and add newline to completion file

joelostblom committed:
Use old format syntax instead of f-strings

joelostblom commented on watson/cli.py:
```diff
@@ -1067,19 +1125,22 @@ def _final_print(lines):
-            )
-            for frame in frames
-        ))
+        for frame in frames:
```
I was not sure how to incorporate the if-statement in this section, so this is the only place in the merge where I went with the older syntax with the traditional for loop. Let me know if you prefer the have this change reverse and how it would be written with the 'note' if statement below.

joelostblom commented on watson/utils.py:
```diff
@@ -420,3 +423,10 @@ def json_arrow_encoder(obj):
+def format_note(note):
+    return u"{}{}".format(
+        style('note', '>> '),
+        style('note', note.replace('\n', '\n' + ' '*20))
```
This aligns rows of notes containing `\n` in the log output.

joelostblom commented on watson/watson.py:
```diff
@@ -302,9 +305,16 @@ def stop(self, stop_at=None):
+        if note is None:
+            note = old.get('note')
+        elif old.get('note') is not None:
+            print('Overwriting old note:\n>> {}'.format(old.get('note')))
```
This means that this warning is printed first, before the always printed watson stop message. This is a bit visually jarring, but it might be a good thing since it is an important thing to note, see screenshot. I tried passing a variable up to watson cli and then use echo after the stop message, but could not find a good way of doing this since the frameattributesare actually properties are note modifieable so I could not pass say a tuple and then print the old note and overwrite the property with the new note.

![image](https://user-images.githubusercontent.com/4560057/85078526-73ca0500-b179-11ea-8a5a-87866cc83c4f.png)

joelostblom committed:
Pluralize consistently

joelostblom committed:
Do not print empty notes after edits and in the log

joelostblom committed:
Remove seemingly redundant note assignment

joelostblom commented:
@k4nar Just a reminder to have a look at this when you get a chance.

joelostblom commented:
@k4nar and @jmaupetit I have been running this branch locally since opening the PR and it seems to be working well (although I don't use notes that often tbh). Just a reminder to review this PR when you have time.

aero31aero reviewed (approved):
Just a reminder, this is a very useful feature imo and should be merged. I looked at the code and it looks fine to me. Tested the branch locally and it works as expected.

Except for maybe the rebase artifacts here, this LGTM.

joelostblom commented:
Just a note that I didn't use this feature much and I am no longer running this branch locally. I ran into some issues with my frames file format when I moved it to a new watson installed from the main branch, so I went ahead and deleted the notes fields for each frame and it is now working again. I didn't have time to investigate exactly why this issue occurred, but thought I would leave a note that  there might be incompatibilities with newer watson versions in case others are considering running from this PR (you probably should wait until when/if this feature is merged into main).
